ballerina/tests/from_json_test.bal

```diff
@@ -1264,3 +1264,350 @@ isolated function testRecordWithRestAsExpectedTypeForParseAsTypeNegative() {
+    readonly int id;
+    readonly ReadonlyFieldsRec3 userDetails;
+};
+
+type ReadonlyFieldsRec9 record {|
+    readonly int id;
+    string name;
+|};
+
+type ReadonlyFieldsRec10 record {
+    readonly int id;
+    string name;
+};
+
+@test:Config
+function testReadonlyFields() returns error? {
+    json user = {"id": 4012, "name": "John Doe"};
+    json user2 = {"id": 4012, "userDetails": user, age: 13};
+
+    ReadonlyFieldsRec1 r1 = check parseAsType(user);
+    test:assertEquals(r1, {id: 4012});
+    test:assertEquals(r1.id, 4012);
+    
+    ReadonlyFieldsRec2 r2 = check parseAsType(user);
+    test:assertEquals(r2, {id: 4012, name: "John Doe"});
+    test:assertEquals(r2.id, 4012);
+    
+    ReadonlyFieldsRec3 r3 = check parseAsType(user);
+    test:assertEquals(r3, {id: 4012, name: "John Doe"});
+    test:assertEquals(r3.id, 4012);
+
+    ReadonlyFieldsRec4 r4 = check parseAsType(user);
+    test:assertEquals(r4, {id: 4012, name: "John Doe"});
+    test:assertEquals(r4.id, 4012);
+
+    ReadonlyFieldsRec5 r5 = check parseAsType(user2);
+    test:assertEquals(r5, {id: 4012, userDetails: {id: 4012, name: "John Doe"}});
+    test:assertEquals(r5.id, 4012);
+    test:assertEquals(r5.userDetails, {id: 4012, name: "John Doe"});
+    
+    ReadonlyFieldsRec6 r6 = check parseAsType(user2);
+    test:assertEquals(r6, {id: 4012, userDetails: {id: 4012, name: "John Doe"}, age: 13});
+    test:assertEquals(r6.id, 4012);
+    
+    ReadonlyFieldsRec7 r7 = check parseAsType(user2);
+    test:assertEquals(r7, {id: 4012, userDetails: {id: 4012, name: "John Doe"}});
+    test:assertEquals(r7.id, 4012);
+
+    ReadonlyFieldsRec8 r8 = check parseAsType(user2);
+    test:assertEquals(r8, {id: 4012, userDetails: {id: 4012, name: "John Doe"}, age: 13});
+    test:assertEquals(r8.id, 4012);
+
+    ReadonlyFieldsRec9 r9 = check parseAsType(user);
+    test:assertEquals(r9, {id: 4012, name: "John Doe"});
+    test:assertEquals(r9.id, 4012);
+    
+    ReadonlyFieldsRec10 r10 = check parseAsType(user);
+    test:assertEquals(r10, {id: 4012, name: "John Doe"});
+    test:assertEquals(r10.id, 4012);
+}
+
+type ReadonlyFieldsRec11 record {|
+    readonly int id;
+    readonly string taxNo = "N/A";
+|};
+
+type ReadonlyFieldsRec12 record {
+    readonly int id;
+    readonly string taxNo = "N/A";
+};
+
+type ReadonlyFieldsRec13 record {|
+    readonly int id;
+    readonly string name;
+    readonly string taxNo = "N/A";
+|};
+
+type ReadonlyFieldsRec14 record {
+    readonly int id;
+    readonly string name;
+    readonly string taxNo = "N/A";
+};
+
+type ReadonlyFieldsRec15 record {|
+    readonly int id;
+    readonly ReadonlyFieldsRec4 userDetails;
+    readonly string address = "N/A";
+|};
+
+type ReadonlyFieldsRec16 record {
+    readonly int id;
+    readonly ReadonlyFieldsRec4 userDetails;
+    readonly string address = "N/A";
+};
+
+type ReadonlyFieldsRec17 record {|
+    readonly int id;
+    readonly ReadonlyFieldsRec3 userDetails;
```

Comment:
Here `ReadonlyFieldsRec3` is immutable because it is a closed record and all the fields are readonly fields. Can we add a test where the type of one field is selectively immutable? E.g., `int[]` so `readonly int[] f;`.